### PR TITLE
Adding a class to simulate spectra from a real background

### PR DESCRIPTION
**Pull request by AtreyeeS**
Hello,

After discussions with @registerrier and @bkhelifi , I have added a class which can simulate spectra from real HESS runs.

I have put in a simple tutorial, but am unsure as to what to put in as test.
Since this code requires real source runs as input, and the numpy random number generator, the results might vary...

One option might be to simulate a 100 crab like runs on the gammapy-extra dataset and see that the average flux and spectral index match the expected value within some tolerance level.

Or we can put in very basic tests that the code works, like opens the files properly... But that seems to be quite elementary.

@cdeil @joleroi @leajouvin : Can you please suggest some tests?

### CODE REVIEW

**Comment by cdeil** [comment]
@AtreyeeS - Thanks!

Did you see the ones we already have?

* http://docs.gammapy.org/dev/notebooks/spectrum_simulation.html (I think by @joleroi)
* http://docs.gammapy.org/dev/notebooks/cta_simulation.html (I think by @jjlk)

Ideally we would have one way to simulate 1D spectra, and have that be flexible to accomodate all use cases. Having two or three is confusing for users and very hard to maintain (each has it's own set of bugs and missing tests / features / docs).

I realise improving existing code is much harder than adding new one, but we have to do that extra effort in Gammapy now.

What does this mean concretely? I don't know, I didn't compare the 2 existing classes and this new one yet. @AtreyeeS or @joleroi or @registerrier - if you can do that, and find a good solution, great! If it's not clear to you either, I can have a look next week and make a suggestion, or we have a quick chat on Skype.

**Comment by AtreyeeS** [comment]
Thanks @cdeil !

I completely agree with you in the sense it will be confusing to have too many classes for spectral simulations.

In fact, my class draws heavily from the spectral simulation class (http://docs.gammapy.org/dev/notebooks/spectrum_simulation.html)

I did intend to merge them, but could not really figure out a non-messy way for it.

It would be really nice to have your inputs!

**Comment by cdeil** [comment]
@AtreyeeS - I have very little time this week, so I won't really be able to contribute to the solution here. Hoping that @joleroi or @registerrier will give feedback.

I did have a quick look at your spectrum simulator, and don't really understand yet why the current [gammapy.spectrum.SpectrumSimulation](http://docs.gammapy.org/dev/api/gammapy.spectrum.SpectrumSimulation.html) shouldn't work for your use case with minor additions / modifications.

You do reflected regions. That seems very specific, and should be separate from 1D spectrum simulation, no? What if I want to estimate the background from the ring or a template background spectrum that was previously estimated from off runs. We don't want to add another spectrum simulator for those different background estimates, right?

You keep track of multiple observations on the simulator object. The exiting `SpectrumSimulation` class focuses on simulating one observation, while providing an API that let's the user simulate multiple observations. That seems more flexible, no?

In your `simulate_background_counts` I see `np.random.poisson(alpha*nOFF)`. That doesn't seem right. Shouldn't we write the spectrum simulator such that `n_pred_on` and `n_pred_off` are computed and then poisson fluctuated? And allow for multiple use cases by having a flexible API to set `n_pred_off`, e.g. from real background. But keep the spectrum simulation simple, basically Poisson-fluctuate the predicted counts, plus some utilities to compute the predicted counts for common applications?

Overall these spectrum simulation classes are only ~ 50 lines of code, and anyone doing a study will always do ~ 100 or even several ~ 100 lines of code to run their simulations and analyse the results. I don't think we should try to pre-code all common use cases how one can run a 1D spectrum simulation. Instead we should try to keep the `SpectrumSimulation` class in Gamampy simple (and not add new ones), and solve the problem by adding more docs how to do such studies, i.e. your new example notebook should remain, but you should try to re-use the existing class and only do small extensions to that class, keeping the rest of the code such as reflected regions computations if really needed in your example notebook.

File: gammapy/spectrum/simulation_realbkg.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        
+        # Add background to on_vector
+        self.on_vector.data.data += bkg_hist * u.ct
+
+        # Create off vector
+        off_events=bkg_res.off_events.table
+        off_events.remove_rows(idxON)
+        alpha_new=float(a_on)/float(a_off - 1)
+
+        off_counts,edge=np.histogram(off_events["ENERGY"],self.on_vector.energy.bins)
```
Comment (cdeil) [inline gammapy/spectrum/simulation_realbkg.py]: Here you're calling `np.histogram` on event energies.

I think we should try to limit ourselves to "binned" simulators by just calling `np.random.poisson`, in the same way how Sherpa only has the `sherpa.fake` calls that poisson-fluctuate counts from npred vectors.

Long-term we might also get unbinned analysis or simulators (i.e. event samplers) in Gammapy, but those should probably be added as separate new classes, not mixed in with the existing binned analysis / simulation code.

@AtreyeeS - Am I missing something here? Or is re-organising your code possible to allow your use case in this way, i.e. move this `np.histogram` call and event handling into an earlier step, where you do that first in your script and then use the existing binned event simulator?

**Comment by AtreyeeS** [comment]
Hello @cdeil

My aim of making this tool was to have something similar to a "fakeit" in XSPEC ... a science case being the user has extracted a spectrum for a (faint) source, and now wants to understand how good the reconstruction is by simulating many of such spectra in the observed sky background (eg: trying to detect a weak source in the galactic plane; then its better to use the observed background instead of the Field of View ones or ones from blank sky observations)

After looking at your comments, one possible solution might be to extract the background outside the class, and then pass it into the SpectrumSimulation class? This may be possible with a little bit of constructor overloading? I would like to keep the function simulate_background_counts() in the class itself, and not leave it to the user.

Apart from just Poission fluctuating the counts, I am also extracting one reflected region and putting in the counts from that in the source. This won't directly work if the user passes a FoV background, so I am not sure if I can make the class generic for all types of background.
Also, I think it would be complex to use the ring background here. Should we not have to account for the acceptance profile in that case? 

So possible solution: Change the existing SpectrumSimulation user can pass either the reflected background or a background model (or none) as parameters, and re-write my notebook significantly to make the science case self-explanatory. This still won't work for other types of backgrounds though. Does this look good to you?

**Comment by cdeil** [comment]
Yes, I think it would be best to limit the scope of the `SpectrumSimulation` class and never have it do background estimation e.g. with reflected regions. The background is something that you decide or estimate before doing the spectrum simulation.

Currently http://docs.gammapy.org/dev/api/gammapy.spectrum.SpectrumSimulation.html takes a `background_model`, which can be any shape by using e.g. a table model.

Are you trying to write a tool that takes a real dataset, and creates a simulation that matches that?
I don't think you can avoid smoothing or fitting the background to the `n_off` counts, because if you just use `alpha x n_off` as background model, it will already have Poisson noise, and then unless your alpha is super small, you can't be sure you don't get weird effects from Poisson-fluctuating the background twice in your simulation (once from the real `n_off` data, once because you call np.poisson)

So if you agree with that assessment, and are willing to create a background spectral model yourself, the existing class works OK, no?
We could also add one or two convenience methods to the existing `SpectrumSimulation` that estimates a background model for a given `n_off` data, no?

---

Actually, I just now looked at `SpectrumSimulation` and saw that `npred_background` applies AEFF and EDISP to the background model, i.e. treats it as a spectral flux model. I'm not sure this is a good idea. I would suggest to change that to either have background model like in http://gamma-astro-data-formats.readthedocs.io/en/latest/irfs/full_enclosure/bkg/index.html (i.e. a rate) or directly ask the user to put in background model as a counts spectrum, and then to maybe have convenience methods to estimate that from given `n_off` or rate spectra.

But even here I'm not sure if that is a good solution. @joleroi - please chime in. Given that this is a very hard problem how to write a spectrum simulator that's usable for all common use cases, we could even do a short call next week to try and figure it out together?

**Comment by AtreyeeS** [comment]
Hi,

A short discussion will be very good... Does Tuesday (13th March) or Wednesday (14th) morning (say 10:am) suit you @cdeil @joleroi @registerrier ? 

Yes, my aim is to have simulations which reproduce my observed spectra, using the background that I used for the spectral extraction.
So one possible solution is, as you are saying, extract the background, fit it with a functional form, and pass that to the existing SpectralSimulation class. But would it not be better to be able to use the background directly, without approximating it by a functional form? 

I am not sure I understand why alpha needs to be small... alpha*nOFF is the observed number of counts in my background region in one particular observation. A user would typically simulate many runs on this one observation, so nbkg=np.random.poisson(alpha*nOFF) should give me the distribution of the background counts.

**Comment by cdeil** [comment]
For me, Tuesday is no good, We or Thursday morning next week is possible.

> nbkg=np.random.poisson(alpha \times nOFF) should give me the distribution of the background counts.

But `nOFF` is already Poisson-fluctuated, so this simulated background will be double-Poisson fluctuated (in a weird way, mixed with alpha). Who knows what happens there statistically, better avoid it, even if it's quite a lot of work to smooth or fit background models, no?

**Comment by joleroi** [comment]
What's the status here, @AtreyeeS should we come up with a TODO list to cover your use case?

**Comment by AtreyeeS** [comment]
Thanks @joleroi for asking! This work was stalled for sometime.
It wil be good to have a small discussion sometime regarding the best method of implementation, then I can put that in gammapy accordingly.

**Comment by cdeil** [comment]
As discussed in the Gammapy dev call, closing this for now without merging.

We can discuss next week how to add a way to set up simulations that are close to real observations best, possibly via small changes to the existing simulation class and a docs notebook, instead of a new separate class.

**Pull request closed without merging**